deepinv/training/adversarial.py

```diff
@@ -0,0 +1,224 @@
+        """
+        logs = {}
+
+        self.optimizer.G.zero_grad()
+
+        # Evaluate reconstruction network
+        x_net = self.model_inference(y=y, physics=physics)
+
+        # Compute reconstructed measurement
+        y_hat = physics.A(x_net)
+
+        ### Train Generator
+        if train or self.display_losses_eval:
+            loss_total = 0
+            for k, l in enumerate(self.losses):
+                loss = l(
+                    x=x,
+                    x_net=x_net,
+                    y=y,
+                    y_hat=y_hat,
+                    physics=physics,
+                    model=self.model,
+                    D=self.D,
+                )
+                loss_total += loss.mean()
+                if len(self.losses) > 1 and self.verbose_individual_losses:
+                    current_log = (
+                        self.logs_losses_train[k] if train else self.logs_losses_eval[k]
+                    )
+                    current_log.update(loss.detach().cpu().numpy())
+                    cur_loss = current_log.avg
+                    logs[l.__class__.__name__] = cur_loss
+
+            current_log = (
+                self.logs_total_loss_train if train else self.logs_total_loss_eval
+            )
+            current_log.update(loss_total.item())
+
+            logs[f"TotalLoss"] = current_log.avg
+
+        if train:
+            loss_total.backward(retain_graph=True)  # Backward the total generator loss
+
+            norm = self.check_clip_grad()  # Optional gradient clipping
+            if norm is not None:
+                logs["gradient_norm"] = self.check_grad_val.avg
+
+            # Generator optimizer step
+            self.optimizer.G.step()
+
+        ### Train Discriminator
+        for _ in range(self.step_ratio_D):
+            if train or self.display_losses_eval:
+                self.optimizer.D.zero_grad()
+
+                loss_total_d = 0
+                for k, l in enumerate(self.losses_d):
+                    loss = l(
+                        x=x,
+                        x_net=x_net,
+                        y=y,
+                        y_hat=y_hat,
+                        physics=physics,
+                        model=self.model,
+                        D=self.D,
+                    )
+                    loss_total_d += loss.mean()
+                    if len(self.losses_d) > 1 and self.verbose_individual_losses:
+                        current_log = (
+                            self.logs_losses_train[k + len(self.losses)]
+                            if train
+                            else self.logs_losses_eval[k + len(self.losses)]
+                        )
+                        current_log.update(loss.detach().cpu().numpy())
+                        cur_loss = current_log.avg
+                        logs[l.__class__.__name__] = cur_loss
+
+            if train:
+                loss_total_d.backward()
+
+                self.clip_grad_D()
```

Comment:
Maybe we can also visualize discriminator norm in the same way that we do for the generator ?